### ISSUE TITLE
Add x, y, w, h properties to Rect2

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#1991]
### Describe the project you are working on
Project using Rect2s.

### Describe the problem or limitation you are having in your project
I happen to use Rect2 quite a lot in my project and it's starting to feel inconvenient and verbose. The reason is that I often have to access the `position` and `size` properties, which results in `rect.position.x` etc. all over the code.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Rect2 should have `x`, `y`, `w`, `h` properties, so you can do e.g. `rect.x` instead of `rect.position.x` or `rect.w` instead of `rect.size.x` etc. It makes code much more compact and arguably easier to read.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
See above.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
The point is to make the lines shorter.

### Is there a reason why this should be core and not an add-on in the asset library?
Impossible without modifying the engine.

**Comment by dalexeev** [comment]
1\. The following properties can also be useful:

```gdscript
rect.lt == rect.position
rect.rt == rect.position + Vector2(rect.size.x, 0)
rect.lb == rect.position + Vector2(0, rect.size.y)
rect.rb == rect.position + rect.size
```

2\. By the way, there are properties for scalar components, and for vector components there are no shorthand. You either need to construct a vector from the desired scalar component of the original vector and zero, or use the `project()` method.

```gdscript
vector.x_vec == Vector2(vector.x, 0) == vector.project(Vector2.RIGHT)
vector.y_vec == Vector2(0, vector.y) == vector.project(Vector2.DOWN)
```

> ![](https://user-images.githubusercontent.com/47700418/102370555-3e8a5c80-3fce-11eb-9786-e40add56422f.png)
>
>where **a**<sub>1</sub>, **a**<sub>2</sub>, **a**<sub>3</sub> are called the **vector components** (or **vector projections**) of **a** on the basis vectors or, equivalently, on the corresponding Cartesian axes _x_, _y_, and _z_ (see figure), while _a_<sub>1</sub>, _a_<sub>2</sub>, _a_<sub>3</sub> are the respective scalar components (or scalar projections). — https://en.wikipedia.org/wiki/Euclidean_vector

**Comment by aaronfranke** [comment]
Makes sense, and [Plane already does this](https://docs.godotengine.org/en/latest/classes/class_plane.html). If added, should be added for AABB too (`x`/`y`/`z` and `w`/`h`/`d`). Also, `width`/`height`/`depth` are worth considering, but both seem perfectly fine to me.

@dalexeev I don't see the problem with `Vector2(vector.x, 0)`, it's not very long and it makes perfect sense. Also, not really related, but I hate the picture you posted, because the coloring doesn't match the typical convention of X,Y,Z = R,G,B.

**Comment by dalexeev** [comment]
@aaronfranke What do you think about the first point? As for the second, there are many ways to achieve the same thing, it just seemed to me that the one I proposed is a little prettier. For comparison:

```gdscript
var lt = rect.position
var rt = rect.position + Vector2(rect.size.x, 0)
var lb = rect.position + Vector2(0, rect.size.y)
var rb = rect.position + rect.size
```

```gdscript
var lt = rect.position
var rt = rect.position + rect.size.x_vec
var lb = rect.position + rect.size.y_vec
var rb = rect.position + rect.size
```

Perhaps there is a better name? `{x,y}_part`, `{x,y}_projection`  or `project_onto_{x,y}()`?

<details>

> I hate the picture you posted, because the coloring doesn't match the typical convention of X,Y,Z = R,G,B.

The picture is not mine, but I can easily fix it:

![](https://user-images.githubusercontent.com/47700418/102489207-34c03200-407e-11eb-8342-24ba440c6ceb.png)

</details>

**Comment by aaronfranke** [comment]
@dalexeev I don't think it's needed. Two of them are redundant (for rb you can use `rect.end`).

**Comment by dalexeev** [comment]
@aaronfranke I would suggest removing the `position` and `end` properties (but I see no problem with synonymous properties) and introducing 4 properties for all **corners**.

![](https://user-images.githubusercontent.com/47700418/102542822-6a3b3e80-40c3-11eb-9258-77616e83443b.png)

Moreover, `position` – `end` sounds heterogeneous, it was more logical to `start` – `end`.

**Comment by timothyqiu** [comment]
If we have both `position` and `begin`, I think it's natural to make the setters different:

* Setting `position` moves the rectangle (keeps `size`)
* Setting `begin` resizes the rectangle (keeps `end`)
